### PR TITLE
Refactor build properties for Linux and OSX builds

### PR DESCRIPTION
**Pull request by ellismg**
Previously, we used the catch all OS=Unix property in MSBuild to
determine if we were building for Windows vs Unix, and we shared the
same MSIL between Linux and OSX.  However, we have already started to
take dependencies on Linux specific APIs and we plan on using OSX
specific APIs very soon.

We don't want to try to ship MSIL that contains both Linux and OSX
specific implementations of a feature, so the build system needs to
change a little.

We'll continue to use the OS property as the entry point for the build
to specificy what platform you want to build for.  However, we now have
three platforms:

1. Windows_NT
2. Linux
3. OSX

However, the value of the OS property should not be tested directly,
instead we introduce properties about the target platform itself:

1. TargetsWindows
2. TargetsUnix
3. TargetsLinux
4. TargetsOSX

Code which works the same across Linux and OSX should be guarded by
TargetsUnix (which is true when OS is either 'Linux' or 'OSX').  Code
specific to one OS can be guarded with either TargetsLinux or
TargetsUnix.

In this change, I have simply rewritten all the '$(OS)' == 'Unix'
conditions to '$(TargetsUnix)' == 'true'.  A few libraries may do Linux
specific stuff, but we will clean up that debt when we go to flesh out
the OSX implementations.

### CODE REVIEW

**Comment by ellismg** [comment]
@stephentoub Can you take a look?

/cc @sokket: You'll want to consume these changes in your upcoming PR and adjust your properties accordingly.
/cc @chcosta: Once merge the output folders for Linux and Mac will be changing.  I am not sure if that will impact your automation or not.

File: dir.props [hunk 1]
```diff
@@ -107,6 +109,15 @@
     <PackagesBasePath Condition="'$(PackagesBasePath)'==''">$(BinDir)$(OSPlatformConfig)</PackagesBasePath>
   </PropertyGroup>
 
+  <!-- Setup common target properties that we use to conditionally include sources -->
+  <PropertyGroup>
+    <TargetsWindows Condition="'$(OS)' == 'Windows_NT'">true</TargetsWindows>
+    <TargetsLinux Condition="'$(OS)' == 'Linux' or '$(OS)' == 'Unix'">true</TargetsLinux>
```
Comment (stephentoub) [inline dir.props]: Consider adding a similar comment here as you have on L64

Comment (ellismg) [inline dir.props]: Will do.

**Comment by stephentoub** [comment]
Looks great, Matt.  Thanks for doing this.

**Comment by weshaggard** [comment]
@ellismg could you please add some information about this somewhere in our wiki as well?

**Comment by ellismg** [comment]
@weshaggard I added information to https://github.com/dotnet/corefx/wiki/Structuring-Interop-Code in a "Build System" section.  This seemed to be the best place.  Please take a look and update or give me feedback.

**Comment by weshaggard** [comment]
Thanks @ellismg I made one tweak but otherwise it looks good.

**Comment by ellismg** [comment]
FYI @akoeplinger, we are going to be moving away from OS=Unix being a valid OS configuration in a few days (I know this is the default for XBuild on Linux and OSX).  You can start using the new OS values (Linux and OSX) today and we'll stop support Unix early next week after we update all our jobs.

**Comment by akoeplinger** [comment]
@ellismg Thanks for the heads up, I've already adapted my build scripts :)

**Pull request merged**